File: Android/scenes/show/src/main/java/io/agora/scene/show/service/rtmsync/InvitationService.kt
package io.agora.scene.show.service.rtmsync

import androidx.annotation.IntDef
import io.agora.rtmsyncmanager.SyncManager
import io.agora.rtmsyncmanager.utils.AUILogger
import io.agora.rtmsyncmanager.utils.GsonTools
import io.agora.rtmsyncmanager.utils.ObservableHelper
import java.util.UUID

class InvitationService(
    private val channelName: String,
    private val syncManager: SyncManager,
    private val interactionService: InteractionService
) {
    private val tag = "InvitationService($channelName)"

    private val key = "invitation"

    private val observerHelper = ObservableHelper<(InvitationInfo) -> Unit>()

    private val messageRetainer = MessageRetainer(syncManager.rtmManager, channelName, key).apply {
        subscribe { msg ->
            val invitationInfo = GsonTools.toBeanSafely(
                msg.content,
                InvitationInfo::class.java
            )
            AUILogger.logger().d(tag, "onInvitationUpdated: $invitationInfo")
            if (invitationInfo != null) {
                if(invitationInfo.type != InvitationType.INVITING){
                    removeMessages{ it.publisherId == msg.publisherId }
                }
                observerHelper.notifyEventHandlers {
                    it.invoke(invitationInfo)
                }
            }
        }
    }

    fun sendInvitation(
        userId: String,
        success: ((InvitationInfo) -> Unit)? = null,
        failure: ((Throwable) -> Unit)? = null
    ) {
        AUILogger.logger().d(tag, "sendInvitation userId: $userId")
        val scene = syncManager.getScene(channelName)
        if (scene == null) {
            AUILogger.logger().d(tag, "scene is null")
            failure?.invoke(Throwable("scene is null"))
            return
        }

        val userInfo = scene.userService.getUserInfo(userId)
        if (userInfo == null) {
            AUILogger.logger().d(tag, "user info is null")
            failure?.invoke(Throwable("user info is null"))
            return
        }

        val invitationInfo = InvitationInfo(
            userId = userId,
            userName = userInfo.userName,
            type = InvitationType.INVITING
        )
        messageRetainer.sendMessage(
            GsonTools.beanToString(invitationInfo) ?: "",
            userId,
            success = {
                AUILogger.logger().d(tag, "sendInvitation success")
                success?.invoke(invitationInfo)
            },
            error = {
                AUILogger.logger().d(tag, "sendInvitation error: $it")
                failure?.invoke(it)
            }
        )
    }

    fun acceptInvitation(
        invitationId: String,
        success: (() -> Unit)? = null,
        failure: ((Throwable) -> Unit)? = null
    ) {
        AUILogger.logger().d(tag, "acceptInvitation invitationId: $invitationId")
        val inviteMessage = messageRetainer.getMessage {
            GsonTools.toBeanSafely(
                it.content,
                InvitationInfo::class.java
            )?.id == invitationId
        }

        if (inviteMessage == null) {
            AUILogger.logger().d(tag, "invitation message is null")
            failure?.invoke(Throwable("invitation message is null"))
            return
        }

        var invitationInfo = GsonTools.toBeanSafely(
            inviteMessage.content,
            InvitationInfo::class.java
        )
        if (invitationInfo == null) {
            AUILogger.logger().d(tag, "invitation info is null")
            failure?.invoke(Throwable("invitation info is null"))
            return
        }

        invitationInfo = invitationInfo.copy(type = InvitationType.ACCEPT)
        messageRetainer.sendMessage(
            GsonTools.beanToString(invitationInfo) ?: "",
            inviteMessage.publisherId,
            success = {
                AUILogger.logger().d(tag, "acceptInvitation success")
                messageRetainer.removeMessages{ it.publisherId == inviteMessage.publisherId }
            },
            error = {
                AUILogger.logger().d(tag, "acceptInvitation error: $it")
            }
        )
        interactionService.startLinkingInteraction(
            userId = invitationInfo.userId,
            success = {
                AUILogger.logger().d(tag, "startLinkingInteraction success")
                success?.invoke()
            },
            failure = {
                AUILogger.logger().d(tag, "startLinkingInteraction error: $it")
                failure?.invoke(it)
            }
        )
    }

    fun rejectInvitation(
        invitationId: String,
        success: ((InvitationInfo) -> Unit)? = null,
        failure: ((Throwable) -> Unit)? = null
    ) {
        AUILogger.logger().d(tag, "rejectInvitation invitationId: $invitationId")
        val inviteMessage = messageRetainer.getMessage {
            GsonTools.toBeanSafely(
                it.content,
                InvitationInfo::class.java
            )?.id == invitationId
        }

        if (inviteMessage == null) {
            AUILogger.logger().d(tag, "invitation message is null")
            failure?.invoke(Throwable("invitation message is null"))
            return
        }

        var invitationInfo = GsonTools.toBeanSafely(
            inviteMessage.content,
            InvitationInfo::class.java
        )
        if (invitationInfo == null) {
            AUILogger.logger().d(tag, "invitation info is null")
            failure?.invoke(Throwable("invitation info is null"))
            return
        }

        invitationInfo = invitationInfo.copy(type = InvitationType.ACCEPT)
        messageRetainer.sendMessage(
            GsonTools.beanToString(invitationInfo) ?: "",
            invitationInfo.userId,
            success = {
                AUILogger.logger().d(tag, "rejectInvitation success")
                messageRetainer.removeMessages{ it.publisherId == inviteMessage.publisherId }
                success?.invoke(invitationInfo)
            },
            error = {
                AUILogger.logger().d(tag, "rejectInvitation error: $it")
                failure?.invoke(it)
            }
        )
    }

    fun subscribe(onUpdate: (InvitationInfo) -> Unit) {
        AUILogger.logger().d(tag, "subscribe onUpdate=$onUpdate")
        observerHelper.subscribeEvent(onUpdate)
    }

    fun unSubscribe(onUpdate: (InvitationInfo) -> Unit) {
        AUILogger.logger().d(tag, "unSubscribe onUpdate=$onUpdate")
        observerHelper.unSubscribeEvent(onUpdate)
    }

    fun release() {
        AUILogger.logger().d(tag, "release")
        observerHelper.unSubscribeAll()
        messageRetainer.release()
    }

}

@IntDef(InvitationType.INVITING, InvitationType.ACCEPT, InvitationType.REJECT)
@Target(AnnotationTarget.FIELD)
@Retention(AnnotationRetention.SOURCE)
annotation class InvitationType {
    companion object {
        const val INVITING = 0
        const val ACCEPT = 1
        const val REJECT = 2
    }
}

data class InvitationInfo(
    val id: String = UUID.randomUUID().toString(),
    val userId: String,
    val userName: String,
    @InvitationType val type: Int,
)

File: iOS/AgoraEntScenarios/Common/API/RTMSyncManager/RTMSyncManager/Classes/Extension/ApplyAndInvitation/InvitationService.swift
//
//  InvitationService.swift
//  RTMSyncManager
//
//  Created by wushengtao on 2024/6/7.
//

import Foundation

private let key = "invitation"

@objc public enum InvitationType: Int, Codable {
    case inviting = 0
    case accept = 1
    case reject = 2
}

@objc public class InvitationInfo: NSObject, Codable {
    public var id: String = UUID().uuidString
    public var userId: String = ""
    public var userName: String = ""
    public var type: InvitationType = .inviting
    
    enum CodingKeys: String, CodingKey {
        case id, userId, userName, type
    }
}

@objc public protocol InvitationServiceProtocol: NSObjectProtocol {
    func onInvitationDidReceive(channelName: String, info: InvitationInfo)
}

public class InvitationService: NSObject {
    
    private var channelName: String
    private var syncManager: AUISyncManager
    private var interactionService: InteractionService
    
    private var respDelegates = NSHashTable<InvitationServiceProtocol>.weakObjects()
    
    private lazy var messageManager: InviteMessageManager = {
        let manager = InviteMessageManager(channelName: channelName, key: key, rtmManager: syncManager.rtmManager)
        return manager
    }()
    
    deinit {
        aui_info("deinit InvitationService[\(channelName)]", tag: "InvitationService")
        innerUnsubscribe()
    }
    
    required init(channelName: String, 
                  syncManager: AUISyncManager,
                  interactionService: InteractionService) {
        aui_info("init InvitationService[\(channelName)]", tag: "InvitationService")
        self.channelName = channelName
        self.syncManager = syncManager
        self.interactionService = interactionService
        super.init()
        innerSubscribe()
    }
    
    private func innerSubscribe() {
        messageManager.subscribe(delegate: self)
    }
    
    private func innerUnsubscribe() {
        messageManager.unsubscribe(delegate: self)
    }
    
    private func createInviteInfo(userId: String, type: InvitationType) -> InvitationInfo? {
        guard let userList = syncManager.getScene(channelName: channelName)?.userService.userList,
              let user = userList.first(where: { $0.userId == userId}) else {
            return nil
        }
        
        let invitInfo = InvitationInfo()
        invitInfo.type = type
        invitInfo.userId = userId
        invitInfo.userName = user.userName
        
        return invitInfo
    }
    
    private func getInviteMessage(invitationId: String) -> InviteMessageInfo? {
        let message = messageManager.getMessage({ $0.content.contains(invitationId)})
        
        return message
    }
    
    private func createInviteInfo(messageInfo: InviteMessageInfo, type: InvitationType) -> InvitationInfo?  {
        let invitInfo: InvitationInfo? = decodeModel(jsonStr: messageInfo.content)
        invitInfo?.type = type
        
        return invitInfo
    }
}

//MARK: public
extension InvitationService {
    public func subscribe(delegate: InvitationServiceProtocol) {
        if respDelegates.contains(delegate) {
            return
        }
        
        respDelegates.add(delegate)
    }
    
    public func unsubscribe(delegate: InvitationServiceProtocol) {
        respDelegates.remove(delegate)
    }
    
    public func sendInvitation(userId: String, completion: ((NSError?)->())?) {
        guard let invitInfo = createInviteInfo(userId: userId, type: .inviting),
              let content = encodeModelToJsonStr(invitInfo) else {
            aui_info("sendInvitation userId: \(userId)", tag: "InvitationService")
            completion?(NSError(domain: "sendInvitation fail", code: -1, userInfo: nil))
            return
        }
        
        aui_info("sendInvitation userId: \(userId)", tag: "InvitationService")
        messageManager.sendMessage(content: content, userId: userId) { err in
            aui_info("sendInvitation userId: \(userId) completion: \(err?.localizedDescription ?? "success")", tag: "InvitationService")
            completion?(err)
        }
    }
    
    public func acceptInvitation(invitationId: String, completion: ((NSError?)->())?) {
        guard let invitMessage = getInviteMessage(invitationId: invitationId),
              let invitInfo = createInviteInfo(messageInfo: invitMessage, type: .accept),
              let content = encodeModelToJsonStr(invitInfo) else {
            aui_info("acceptInvitation invitationId: \(invitationId) fail", tag: "InvitationService")
            completion?(NSError(domain: "acceptInvitation fail", code: -1, userInfo: nil))
            return
        }
        
        aui_info("acceptInvitation invitationId: \(invitationId)", tag: "InvitationService")
        messageManager.sendMessage(content: content, userId: invitMessage.publisherId) {[weak self] err in
            aui_info("acceptInvitation invitationId: \(invitationId) completion: \(err?.localizedDescription ?? "success")", tag: "InvitationService")
            if err == nil {
                self?.messageManager.removeMessages(filter: { $0.publisherId == invitMessage.publisherId})
            }
            completion?(err)
        }
        
        interactionService.startLinkingInteraction(userId: invitInfo.userId) { err in
            aui_info("startLinkingInteraction: \(invitationId) completion: \(err?.localizedDescription ?? "success")", tag: "InvitationService")
        }
    }
    
    public func rejectInvitation(invitationId: String, completion: ((NSError?)->())?) {
        guard let invitMessage = getInviteMessage(invitationId: invitationId),
              let invitInfo = createInviteInfo(messageInfo: invitMessage, type: .accept),
              let content = encodeModelToJsonStr(invitInfo)  else {
            aui_info("rejectInvitation invitationId: \(invitationId) fail", tag: "InvitationService")
            completion?(NSError(domain: "rejectInvitation fail", code: -1, userInfo: nil))
            return
        }

        aui_info("rejectInvitation invitationId: \(invitationId)", tag: "InvitationService")
        messageManager.sendMessage(content: content, userId:  invitMessage.publisherId) {[weak self] err in
            aui_info("rejectInvitation invitationId: \(invitationId) completion: \(err?.localizedDescription ?? "success")", tag: "InvitationService")
            if err == nil {
                self?.messageManager.removeMessages(filter: { $0.publisherId == invitMessage.publisherId})
            }
            completion?(err)
        }
    }
}

//MARK: InviteMessageProtocol
extension InvitationService: InviteMessageProtocol {
    public func onNewInviteDidReceived(channelName: String, message: InviteMessageInfo) {
        guard let info: InvitationInfo = decodeModel(jsonStr: message.content) else { return }
        aui_info("onNewInviteDidReceived: \(message.content)", tag: "InvitationService")
        if info.type != .inviting {
            messageManager.removeMessages { $0.publisherId == message.publisherId}
        }
        respDelegates.allObjects.forEach { delegate in
            delegate.onInvitationDidReceive(channelName: channelName, info: info)
        }
    }
}
